File: Java/Heaps/Heap_array_basic.java
// Implmenting Heap Data structure in Java: 4th October 2017:

import java.util.Scanner;
import java.util.ArrayList;

// Rules/invariants about a Heap Structure: Implemented using an array:
// 1. If a node has index 'i' , then its left child is '2i + 1' and right child is '2i+2'
// 2. If a node has index 'i', then the index of its parent is

// Equal keys not allowed in this heap:

// Create a class for the Heap Data structure:
class Heap {

private ArrayList<Double> heapArray;
private int curr_index;


public Heap(){

  heapArray = new ArrayList<Double>(1);
  curr_index = -1;

}

// Methods for the Heap Data structure:

// 0. Method to swap two elements of an array:
public void swap(int ind_1, int ind_2){

  double temp = heapArray[ind_1];
  heapArray[ind_1] = heapArray[ind_2];
  heapArray[ind_2] = temp;

}

// 1. Method to bubble up the last element: Reheapification upwards:
// Recursive function:
public void bubble_up(int bubble_index){

  // Trivial case: Element is at the root or it is less than its parent:
  if (heapArray[bubble_index] < heapArray[(int) (bubble_index-1)/2] || bubble_index ==0)  {
    return;
  }

  // Non-trivial case:
  else {
    swap(bubble_index, (int) (bubble_index-1)/2);
    bubble_up((int) (bubble_index-1)/2);
  }

}

// 2. Method to bubble down the root element: Reheapification downwards:
// Recursive function:
public void bubble_down(int bubble_index){

  // Trivial case: Element is in a leaf node:
  if (curr_index < 2*bubble_index + 1)  {
    return;
  }

  // Non -trivial case: But first check to see if the element at bubble index is already greater than its children:
  else{

      if (heapArray[bubble_index] < heapArray[2*bubble_index +1] || heapArray[bubble_index] < heapArray[2*bubble_index +2] ){

        // Check to see which of the two children are greater:
        if (heapArray[2*bubble_index +1] > heapArray[2*bubble_index +2] ){
          swap(bubble_index, 2*bubble_index + 1);
          bubble_down(2*bubble_index + 1);
        }

        else{
          swap(bubble_index, 2*bubble_index + 2);
          bubble_down(2*bubble_index + 2);
        }

      }

      return;
  }

}


// 3. Method to insert a new key into the heap:
public Boolean insert(double key){

    // Add the new key at the end of the HeapArray:
    heapArray[++curr_index] = key;

    // Bubble up the new key until the heap structure is maintained:
    bubble_up(curr_index);

    return true;  // Put exceptions later on:
}

// 4. Method to remove the root key  from the heap: Like in a Priority Queue:
public double remove_root(){

  // Extract the root element:
  double root = heapArray[0];

  // Shift the last element to the zeroth index and decrement size of heap by 1:
  heapArray[0] = heapArray[curr_index--];

  // Bubble down the zeroth index element until the heap structure is maintained:
  bubble_down(0);

  return root;

}


// 5. Method to heapify an unordered heap:

// 6. Method to print the entries in the heap:
public void print(){

  for (int i=0;i<=curr_index;i++){
    System.out.print(heapArray[i] + ",") ;
  }
    System.out.println(" ");

}

// Main function to test the implementation:
public static void main(String[] args) {

    // Create a new Heap:
    Heap myHeap = new Heap();

    String userEntry;

    // Create a scanner object to get user input:
    Scanner myScan = new Scanner(System.in);

    System.out.println("Please enter the elements/keys of the Heap");

    Boolean stopLoop = false;
    double key;

            while (!stopLoop) {
                userEntry = myScan.nextLine();

                try {

                      key = Double.parseDouble(userEntry);
                      myHeap.insert(key);

                } catch (NumberFormatException e) {

                    if (userEntry.equals("q") || userEntry.equals("Q")) {
                        System.out.println("Stopping loop");
                        stopLoop = true;

                    } else {
                        System.out.println("Please enter a valid character");
                    }
                }

            }   // END while loop

      // Check the order of the elments in the heap:
      System.out.println("Elements of the original Heap are: ");
      myHeap.print();

      // Test the removal of an entry in the heap:
      myHeap.remove_root();


      // Now recheck the entries in the heap:
      System.out.println("Elements of the  Heap with root removed are: ");
      myHeap.print();

}




}  // END class Heap:

File: C++/Heaps/Heap_array_basic.cpp
/*
 * Heap_array_basic.cpp
 *
 *  Created on: Sep 16, 2017
 *      Author: savio
 */

#include <iostream>
#include <math.h>
#include <ctype.h>
#include <stdlib.h>
#include <string>
#include <ctime>

using namespace std;

// A class to implement a max heap using a dynamic array: The entries in the heap structure are integers:
// TO DO: Implement a templated class so that other data types can be used as entries in the heap:
class Heap {

private:

	// Expected size of the heap:
	static const int INITIAL_SIZE = 10;

	// Array to store the heap entries:
	int* heapArray;

	// Current number of entries in the heap:
	int numEntries;

	// Current index of the array:
	int currIndex;

public:
	// 1. Constructor:
	Heap(){
		 heapArray = new int[INITIAL_SIZE];
		 numEntries =0;
		 currIndex =-1;
	}

	// 2. Parameterized constructor:
	Heap(int* inputArray, int length){

		heapArray = new int[INITIAL_SIZE];
		numEntries =0;
		currIndex =-1;

		for (int i=0;i<length;i++){

			heapArray[i]= inputArray[i];
			currIndex++;
		}

	}

	// 3. Using the default destructor of the class itself: ---

	// 4. Method to print the entries in the heap:
	void print(){

		for (int i=0;i<=currIndex;i++){
			cout<<heapArray[i]<<",";
		}

		cout<<endl<<endl;
	}



	// 5. Method to swap entries at two given indices in the array:
	void swap(int index_1 , int index_2){
		 int temp = heapArray[index_1];
		 heapArray[index_1] = heapArray[index_2];
		 heapArray[index_2] = temp;
	}

	// 6. Method to bubble up the last entry in the heap array (Reheapification upwards):
	bool bubble_up(int bubbleIndex){

		// Trivial case: If the entry is already at the root
		if (bubbleIndex ==0){

			return true;
		}


		// Recursive case:
		if (heapArray[bubbleIndex] > heapArray[(int) floor((bubbleIndex- 1)/2)]){

			swap(bubbleIndex , (int) floor((bubbleIndex-1)/2) );

			// Recursively call the bubble up function:
			bubble_up((int) floor((bubbleIndex-1)/2));

		}

		else {
			return true;
		}

	}


	// 7. Method to bubble down the first entry in the heap (Reheapification downwards): Element needs to
	// be swapped with the larger of the two children:
	bool bubble_down(int bubbleIndex){

		// Trivial case: If the entry is already at the leaf nodes
		if (currIndex < 2*bubbleIndex + 1){
			return true;
		}

		else {

			// Case 1: Bubble Index is smaller than its children:
			if (heapArray[bubbleIndex] < heapArray[2*bubbleIndex+1] && heapArray[bubbleIndex] < heapArray[2*bubbleIndex+1] ){

				swap(bubbleIndex, 2*bubbleIndex+1);

				bubble_down(2*bubbleIndex+1);
			}

			if (heapArray[bubbleIndex] < heapArray[2*bubbleIndex+2] && heapArray[bubbleIndex] < heapArray[2*bubbleIndex+2]){

				swap (bubbleIndex, 2*bubbleIndex+2);

				bubble_down(2*bubbleIndex+2);
			}

			return true;
		}
	}

	// 8. Method to add an entry to the heap:
	void add_entry(int entry){

		// First add the entry one after the last index in the heap:
		heapArray[++currIndex] = entry;

		// Now bubble up the last entry in the heap Array:
		bubble_up(currIndex);


	}

	// 9. Method to remove the first  entry from the heap: First remove the head node (first index) entry from
	// the heap. Then replace the first element with the last element. Then bubble down the first
	// element until the heap structure is restored:
	int remove_entry(){

		int res =heapArray[0];

		heapArray[0] = heapArray[currIndex--];

		bubble_down(0);

		return res;
	}


	// 10. Function to construct a heap from a non-heap array:
	void Heapify(){

		for (int i = currIndex; i>=0;i--){
			bubble_down(i);
		}
	}
};




// Main function of the Executable:

int main(int argc, char **argv) {

	// Create an object of a a heap:
	Heap* firstHeap = new Heap();

	// Test out the add entry function:

	// Ask the user for values to enter:
	cout<<" Testing the insert method of the Heap Class ..."<<endl<<endl;
	cout<<" Please input the entries of the Heap . Press 'q' to quit"<<endl;

	string userInput;

	bool stopLoop = false;

	while (!stopLoop){


	getline(cin,userInput);

	try {

		int userNumInput = stoi(userInput);
		firstHeap->add_entry(userNumInput);
	}
	catch (const std::invalid_argument& ia) {


		if (userInput[0] == 'q' || userInput[0]=='Q'){
		stopLoop =true;
	}
	else{
			std::cerr << "Invalid argument: " << ia.what() << '\n';
	}

}

	}


		cout<<" The entries of the heap after insertion are: "<<endl;
		firstHeap->print();

		// Test out the remove entry function:
		int first = firstHeap->remove_entry();

		cout<<" Testing out the Remove method of the Heap Class  for a Priority Queue:"<<endl<<endl;

		cout<<" The First entry removed from the heap is: "<< first<<endl;

		cout<<" The entries in the heap after removing the first entry are"<<endl;
		firstHeap->print();



	// Create a random int array:

	cout<<" Testing out the Heapify method of the Heap class ..."<<endl<<endl;

	int inputArray[11];

	for (int i=0;i<10;i++){
		inputArray[i] = rand() %100 +1;
	}

	//int inputArray[8] = {3,1,4,5,9,2,7,8};

	int length = sizeof(inputArray)/sizeof(inputArray[0]);


	// Second heap: To test the heapify function:
	Heap* secondHeap = new Heap(inputArray, length);


	cout<<" These are elements of a random array of 10 integers between 1 and 100"<<endl;
	secondHeap->print();

	secondHeap->Heapify();

	cout<<" These are the elements of the array after the Heapify operation:"<<endl;
	secondHeap->print();


	return 0;

}
